src/json.cpp

```diff
@@ -1917,57 +2049,61 @@ void TextJsonIn::error( int offset, const std::string &message )
+    const int max_context_chars = 480; // limits context length to this many chars
     err << "\n\n";
     stream->seekg( offset, std::istream::cur );
-    size_t pos = tell();
-    rewind( 3, 240 );
-    size_t startpos = tell();
-    std::string buffer( pos - startpos, '\0' );
-    stream->read( buffer.data(), pos - startpos );
-    auto it = buffer.begin();
-    for( ; it < buffer.end() && ( *it == '\r' || *it == '\n' ); ++it ) {
-        // skip starting newlines
-    }
-    for( ; it < buffer.end(); ++it ) {
-        if( *it == '\r' ) {
-            err << '\n';
-            if( it + 1 < buffer.end() && *( it + 1 ) == '\n' ) {
-                ++it;
+    // remember positions of several places to print a few lines of context
+    const size_t cursor_pos = tell();               // exact position of error
+    rewind( 1, max_context_chars );                 // rewind to start of line
+    const size_t start_of_line = tell();            // start of error line
+    rewind( max_context_lines, max_context_chars ); // rewind a couple lines to show context
+    const size_t start_of_context = tell();         // start of context (couple lines above)
+    size_t end_of_line;                             // end of error line (without \n)
+
+    {
+        seek( cursor_pos );
+        // find end of line and store in endpos
+        end_of_line = tell();
+        while( stream->peek() != EOF ) {
+            if( stream->peek() == '\n' ) {
+                break;
             }
-        } else {
-            err << *it;
-        }
-    }
-    if( !is_whitespace( peek() ) && stream->good() ) {
-        err << peek();
-    }
-    // display a pointer to the position
-    rewind( 1, 240 );
-    startpos = tell();
-    err << '\n';
-    if( pos > startpos ) {
-        err << std::string( pos - startpos, ' ' );
-    }
-    err << "^\n";
-    seek( pos );
-    // if that wasn't the end of the line, continue underneath pointer
-    char ch = stream->get();
-    if( ch == '\r' ) {
-        if( peek() == '\n' ) {
             stream->get();
+            end_of_line = tell();
         }
-    } else if( ch == '\n' ) {
-        // pass
-    } else if( peek() != '\r' && peek() != '\n' && !stream->eof() ) {
-        for( size_t i = 0; i < pos - startpos + 1; ++i ) {
-            err << ' ';
+    }
+    {
+        // print context lines to start of error line
+        seek( start_of_context );
+        std::string buffer( start_of_line - start_of_context, '\0' );
+        stream->read( buffer.data(), start_of_line - start_of_context );
+        err << buffer;
+    }
+    {
+        // print error line
+        std::string buffer( end_of_line - start_of_line, '\0' );
+        stream->read( buffer.data(), end_of_line - start_of_line );
+        err << color_error << buffer << color_end << "\n";
+    }
+    {
+        seek( start_of_line );
+        std::string buffer( cursor_pos - start_of_line, '\0' );
```

Comment:
I think it might be safer to check if `cursor_pos` points to a starting byte of a utf8 sequence and increment it to point to the next starting byte if it doesn't, because otherwise it may lead to incorrect width and out-of-bounds access:

If `cursor_pos` points to the middle of a utf8 sequence here, `buffer` will contain an incomplete utf8 sequence which might cause `utf8_width` to calculate the wrong width. Furthermore, if the incomplete sequence contains only one byte, `UTF8_getch` will try to access the second byte without checking for the buffer length, causing out-of-bounds access (which, on a side note, sounds pretty grim considering the original comment in `catacharset.cpp` suggests it was originally from SDL2... hope they either fixed it or are only using it for c-style strings.)

Reply:
Went with rewinding to start of a codepoint (or stream start) instead - the cursor makes a bit more sense then

TextJsonIn::rewind seems to be a bit borked - `rewind( /* max_lines = -1 */, /* max_chars = 1 */ )` will rewind twice instead of once